### ISSUE TITLE
Possible -addnode bug

### ISSUE THREAD

**Issue by Justaphf** [gandrewstone/bitcoin#65]
I haven't investigated deep enough into the code to really determine if this is an actual bug in the way I suspect, but I as I was going through the code as part of my own personal exploration, I wanted to point this out.  Perhaps someone more familiar with the code can quickly verify if this is actually an issue or not.  If it is, it's a "one-line fix"™.

After a few hours looking into this area of the code, my initial assumptions are the following:
1. Command line arguments plus anything in the bitcoin.conf file only apply to the system once per execution, during startup.  The bitcoin.conf file is not monitored for changes during execution (though this could be quite useful, at least for some key parameters/during testing -- RPC reread-config command FTW).
2. RPC allows manual addition and removal of outbound nodes, but this does not affect the contents of the mapMultiArgs list for the "-addnode" option. **Not sure about this one, haven't looked into the RPC module much yet.**

In net.cpp, function AcceptConnection(...), the computation for nMaxInbound seems semantically incorrect. (NOTE: All comments in below code block are mine.  Only the single line of code is from the original source-code.)

``` c++
    //REVISIT: We are taking the max number of total allowable connections (inbound + outbound),
    //         then subtracting the maximum number of "concurrently allowable" outbound connections (hard coded at 8),
    //         then subtracting the number of "-addnode" parameters (user configured w/o limit checks against spec'd maximum number of outbounds)
    //NOTE: A cursory look through the code doesn't show the number of "-addnode" arguments being decreased.
    //      Given these observations, we could end up in some erroneous cases based on order of operations
    //      and the number of "-addnode" paramters that the user configured via command line/bitcoin.conf.
    //CASE 1:  (BAD, UNLIKELY) I list out an arbitrarily large number of "-addnode" parameters in my bitcoin.conf file 
    //         i.e. every node currently reachable by bitnoes.21.co ~5-6k.  In this case, unless we specify really high -maxconnections
    //         we end up with a negative nMaxInbound value, preventing all inbound connections. (I have not tested this)
    //CASE 2:  (MOST LIKELY) I list a reasonable (though greater than the recommended 8) number of "-addnode" arguments.  The current implementation,
    //         assuming mapMultiArgs doesn't get dynamically updated, will both allow more than the hardcoded default of 8 outbound connections
    //         AND reduce the total number of possible inbound connections.
    //FIX: If my assumptions are correct, this value should actually be:
    //     nMaxInbound = nMaxConnections - MAX_OUTBOUND_CONNECTION (could leave up to 8 connection slots empty if no outbounds connect,
    //                                                              though this seems unlikely with the various seeding methods)
    int nMaxInbound = nMaxConnections - MAX_OUTBOUND_CONNECTIONS - mapMultiArgs["-addnode"].size();
```

In practice, I have run into the issue where I was able to get more than 8 outbounds to connect (it's a feature, not a bug, honest!).  I have also never been able to saturate my -maxconnections, though I'm running on a recently retired gaming rig over a FIOS fiber connection that's basically 4 hops from the Ashburn MCI trunk line, so I always set it pretty high, making the issue less noticeable.  I have a feeling though that some of the "How can I increase the number of connections to my node" posts I've seen where users don't come close to reaching the number of nodes they set as their max, but have gotten past the "your router needs to port-forward" issue have posted about.

**Comment by ptschip** [comment]
Thanks for taking the time to go through the code in detail!  But the calculation for nMaxInbound is correct.  There was a bug in addnode for a long time where if  you were using addnode and then your node you were trying to connect to got recycled, then you wouldn't be able to connect to it again because by the time that node had come back up your outbound slots would be max'd at 8.  So we had to create a second semaphore to handle the addnode connections and therefore you can now have up to 16 outbound slots if you use the maximum of 8 addnode connections and 8 regular.  So to keep the expected behavior of 125 max connections we subtract the addnodes that are in the config file...it's not really a big deal either way IMO but we just put it in there anyway in case anyone noticed they had more than 125 connections.  I really didn't think anybody would look through the code is such detail...great stuff.  Maybe you can help us do some coding?

One thing though, is that maybe you are a bit right here...I think we should still be subtracting the number of addnodes IMO but only to a maximum of 8...because anybody could put any number of addnodes in their config file...any thoughts ?

**Comment by Justaphf** [comment]
Thanks for the quick response.  Based on your mention of using the semaphores to limit the number of concurrent outbound connections, and having several more hours diving into the code, I think I've found a few more related things that seemed like (small) issues, and coded in possible fixes.  At the moment I haven't yet compiled or tested them as I'm a Windows guy and just have the source loaded into Visual Studio 2015.  I need to set up a linux box for doing compiles/testing, and then learn the debugging tools in the linux world.

In the meanwhile I'll commit my changes to my fork so I can pull on a Pine64+ that I already build the release client on and is nearly finished synching.  I can use that to compile and test on.  If you want to look at what I've got so far on my fork https://github.com/BitcoinUnlimited/BitcoinUnlimited/compare/0.12.1bu...Justaphf:0.12.1bu
I'll let you know once I've had a chance to compile and do some testing.

In terms of helping to code, I'd like to help out. My available free time can be pretty unpredictable though and I've spent the last 10 years writing managed code in .NET so I'm still regaining my "C-legs".  Is there a better forum for discussing than a bug report?

Thanks!

**Comment by ptschip** [comment]
if you want to discuss further we hangout on slack...ask @thezerg in the
bitco.in forum for an access to the slack channels

i develop only on windows also...you don't need to setup linux if you
don't want to..you can install mingw which i use, or you can ask
@thezerg i think he uses visual studio.

**Comment by Justaphf** [comment]
Thanks, I joined the channel.  Looks like thezerg doesn't develop BU on MSVC so I'm in the process of trying to get it configured myself while writing up a document for the whole process.

In relation to this issue, I took a little time out from trying to get things set up in MSVC to run some quick test cases on the 12.1 Windows 64-bit release client to prove to myself the things I thought were issues are and that they behave the way I expected.
1. If you don't have any "addnode=" lines in your bitcoin.conf file, then the RPC "addnode <node> add" call is broken.  This is due to the semaphore being initialized based on the number of addnode entries in the config file.  As a result, the RPC addnode calls will never be able to get above min(addnodes in config, MAX_OUTBOUND_CONNECTIONS) since that's the count the semaphore is initialized to (0 in this test case).
2. The RPC "addnode <node> onetry" call does not respect either semaphore, so you could use this to add an unlimited number of outbound nodes, completely ignoring the semaphores.  Since these are outbound connections not tracked in the added nodes list, they also won't affect the number of incoming nodes (though I did not fully verify the inbound connection assertion through testing).
3. Placing more "addnode=" entries in your bitcoin.conf file than "maxconnections=" value blocks all incoming connections.  I tested this by setting my max connections to 50 and adding 60 "addnode=127.0.0.1" lines to my config file.  Placing less than maxconnections but more than 8 also eats up connection slots, though you will still be able to achieve (maxconnections - (2 \* MAX_OUTBOUND_CONNECTIONS)) inbound connections.
   3.a. I know it's user error, but seems like allowing self-referencing IPs should be caught & ignored.
   3.b. Similarly, duplicate IPs should probably also be caught & ignored.
   3.c. Using the "addnode <node> add" RPC call did add to the list as expected, though there was a decent wait while the repetitive failures of the 127.0.0.1 addresses cycled through the wait queue.
   3.d. There doesn't seem to be any validation what-so-ever on the RPC.  Executing "addnode Cr@zy$tr!ng add" adds in "Cr@zy$tr!ng" to the addnodes list.

As an additional note, and I did not directly test this, but based on 3.c. above I believe will behave the way I expect, adding more than 8 valid addnode connections to the bitcoin.conf file could end up re-causing the same bug that the secondary addnode semaphore was added to fix (that all outbound slots would be filled by the time a recycled node came back up).  This is assuming that nothing special happens when a node gets recycled, just goes back onto the wait-queue of the addnode semaphore instead of the original outbound semaphore (I have not looked at node recycling code yet, so not certain how that works).

Suggestions for fixes:
1. Make RPC "addnode <node> onetry" respect the addnode semaphore.  I suggest implementing as a try-wait call so that if there are already max outbound addnode connections it just throws it out without attempting to connect (with appropriate error message back through the RPC call about all outbound connection slots being filled).
1. For issues 1 and 3, I believe the fix could be the same for both, however I see 2 paths.

First, the simple case would be to always initialize the addnode semaphore to MAX_OUTBOUND_CONNECTIONS.  This ensures that the RPC call isn't broken when no addnode entries are in the bitcoin.conf file and that if you put more than 8 addnode parameters in the bitcoin.conf file you never take more than MAX_OUTBOUND_CONNECTIONS away from incoming connection pool.  This option results in the possibility of permanently wasting up to 8 inbound connection slots if the user never configures any "addnodes", but is a very straightforward code change.

A slightly more complex option would be to try and dynamically set the max number of inbound connections based on the number of addnode connections currently being tracked.  This could be accomplished by always setting the addnodes semaphore to 8 and using the following in the AcceptConnection function

``` c++
    LOCK(cs_vAddedNodes);
    int nMaxAddNodeOutbound = std::min((int)vAddedNodes.size(), MAX_OUTBOUND_CONNECTIONS);
    int nMaxInbound = nMaxConnections - MAX_OUTBOUND_CONNECTIONS - nMaxAddNodeOutbound;
```

However, this runs into 2 issues.  If you have saturated the connection pool with inbound connections, but had specified less than 8 addnodes, then using RPC you do addnode, you could end up exceeding maxconnections.  This case could probably be handled by trying to evict an inbound connection to make room for the outbound (or just let maxconnections be exceeded), but not sure the gain is worth the added complexity.
There is also an additional flaw with this in that using vAddedNodes is only checking to see how many addnodes the user has configured, but ignores the fact that the nodes may not be valid node addresses or even if they are, that they are actually connected.

OK, so fairly long winded, but that's my analysis of the issue and possible solutions.

**Comment by ptschip** [comment]
that's a very good all around analysis and investigation...the only
thing i don't quite agree with is the addnode onetry...i beleive it is
the way it is supposed to work...however i
think your'e right on the other points.  I'm looking forward to seeing
you post some PR's!

if you have trouble getting up on msvc then mingw is a good option and i
can give you an install script for that...it's free and works very well
and also lets you run the python regression scripts (but keeping in mind
, many of them don't work on windows)

**Comment by Justaphf** [comment]
Closed as PR #74 was merged for this issue.
